c++/src/kj/compat/http.c++

```diff
@@ -3489,6 +3489,431 @@ WebSocketPipe newWebSocketPipe() {
+  //
+  // The following issues would prevent a struct from being populated:
+  //  Key issues:
+  //    - `Key` is invalid (see `parseKeyName()`).
+  //    - `Key` is repeated.
+  //  Value issues:
+  //    - Got a `Value` when none was expected (only the `max_window_bits` parameters expect values).
+  //    - Got an empty `Value` (0 characters, or all whitespace characters).
+
+  if (params.size() > 4) {
+    // We expect 4 `Key`s at most, having more implies repeats/invalid keys are present.
+    return nullptr;
+  }
+
+  UnverifiedConfig config;
+
+  for (auto& param : params) {
+    KJ_IF_MAYBE(paramType, parseKeyName(param.key)) {
+      // `Key` is valid, but we still want to check for repeats.
+      const auto& side = paramType->side;
+      const auto& property = paramType->property;
+
+      if (property == ParamType::NO_CONTEXT_TAKEOVER) {
+        auto& takeOverSetting = (side == ParamType::CLIENT) ?
+            config.clientNoContextTakeover : config.serverNoContextTakeover;
+
+        if (takeOverSetting == true) {
+          // This `Key` is a repeat; invalid config.
+          return nullptr;
+        }
+
+        if (param.val != nullptr) {
+          // The `x_no_context_takeover` parameter shouldn't have a value; invalid config.
+          return nullptr;
+        }
+
+        takeOverSetting = true;
+      } else if (property == ParamType::MAX_WINDOW_BITS) {
+        auto& maxBitsSetting =
+            (side == ParamType::CLIENT) ? config.clientMaxWindowBits : config.serverMaxWindowBits;
+
+        if (maxBitsSetting != nullptr) {
+          // This `Key` is a repeat; invalid config.
+          return nullptr;
+        }
+
+        KJ_IF_MAYBE(value, param.val) {
+          stripLeadingSpace(*value);
+          if (value->size() == 0) {
+            // This is equivalent to `x_max_window_bits=`, since we got an "=" we expected a token
+            // to follow.
+            return nullptr;
+          }
+          maxBitsSetting = param.val;
+        } else {
+          // We know we got this `max_window_bits` parameter in a Request/Response, and we also know
+          // that it didn't include an "=" (otherwise the value wouldn't be null).
+          // It's important to retain the information that the parameter was received *without* a
+          // corresponding value, as this may determine whether the offer is valid or not.
+          //
+          // To retain this information, we'll set `maxBitsSetting` to be an empty ArrayPtr so this
+          // can be dealt with properly later.
+          maxBitsSetting = ArrayPtr<const char>();
+        }
+      }
+    } else {
+      // Invalid parameter.
+      return nullptr;
+    }
+  }
+  return kj::mv(config);
+}
+
+inline kj::Maybe<size_t> extractBitsFromValue(ArrayPtr<const char>&& bits) {
+  kj::Maybe<size_t> window;
+  kj::Maybe<kj::Exception> maybeException = kj::runCatchingExceptions([&]() {
```

Comment:
Catching all exceptions like this and assuming the exception must be due to a parse error is not very good form. What if some other assertion fails in here? Also, throwing and catching an exception is very expensive.

Instead, I'd recommend using `strtoul` directly, so that you can check if the parse failed.

Alternatively, maybe we should add `tryParseAs` methods to `StringPtr`, which return a `Maybe` rather than throwing... this has come up a lot.

Reply:
Added `tryParseAs` to `StringPtr`: https://github.com/capnproto/capnproto/pull/1480/commits/181ab0032975fc4db4079ba3c60fd5e9cf04822d.
Also updated this code to use it: https://github.com/capnproto/capnproto/pull/1480/commits/113ae4304be821d5b3479064462a1279d8183f69